store/cachemulti/store.go

```diff
@@ -127,6 +127,36 @@ func (cms Store) Write() {
 	}
 }
 
+// Clone creates a snapshot of each store of the cache-multistore.
+// Each copy is a copy-on-write operation and therefore is very fast.
+func (cms Store) Clone() types.CacheMultiStore {
+	stores := make(map[types.StoreKey]types.CacheKVStore, len(cms.stores))
+	for key, store := range cms.stores {
+		stores[key] = store.Clone()
+	}
+	return Store{
+		db:     cms.db.Clone(),
+		stores: stores,
+
+		keys:         cms.keys,
+		traceWriter:  cms.traceWriter,
+		traceContext: cms.traceContext,
+	}
+}
+
+// Restore restores the cache-multistore cache to a given snapshot.
+func (cms Store) Restore(s types.CacheMultiStore) {
+	ms := s.(Store)
+	cms.db.Restore(ms.db)
+	for key, store := range cms.stores {
+		store.Restore(ms.stores[key])
+	}
+
+	// this multi-store must not be used anymore
+	ms.db = nil
+	ms.stores = nil
+}
```

Comment:
`s` is provided as a value parameter, so setting `ms.db = nil` does not affect the `s` passed to `Restore` by the caller.

Reply:
fixed, no need to clear the original one, because the `Restore` calls on internal stores will clear them now.

Reply:
It's very surprising if a method called Restore will mutate its argument, much less render it unusable.